File: agent/src/SCKAudioCapture.m
#import <Foundation/Foundation.h>
#import <ScreenCaptureKit/ScreenCaptureKit.h>
#import <CoreMedia/CoreMedia.h>

// Expose a C function pointer type for Rust to provide
typedef void (*AudioCallback)(const uint8_t *data, size_t length);

@interface AudioCapture : NSObject <SCStreamOutput>
@property (nonatomic, assign) AudioCallback callback;
@end

@implementation AudioCapture
- (void)stream:(SCStream *)stream didOutputSampleBuffer:(CMSampleBufferRef)sampleBuffer ofType:(SCStreamOutputType)type {
    if (type != SCStreamOutputTypeAudio) return;
    
    CMBlockBufferRef blockBuffer = CMSampleBufferGetDataBuffer(sampleBuffer);
    if (!blockBuffer) return;
    
    size_t length = 0;
    char *ptr = NULL;
    
    OSStatus status = CMBlockBufferGetDataPointer(blockBuffer, 0, NULL, &length, &ptr);
    if (status == kCMBlockBufferNoErr && ptr && self.callback) {
        // Send data directly to Rust callback!
        self.callback((const uint8_t *)ptr, length);
    }
}
@end

// Keep strong references so SCK objects aren't deallocated!
static SCStream *globalStream = nil;
static AudioCapture *globalCapture = nil;

// Expose C API to Rust
void start_sck_capture(AudioCallback callback) {
    if (@available(macOS 12.3, *)) {
        [SCShareableContent getShareableContentExcludingDesktopWindows:NO onScreenWindowsOnly:YES completionHandler:^(SCShareableContent *content, NSError *error) {
            
            if (error || !content.displays.firstObject) {
                fprintf(stderr, "SCK Error: %s\n", error ? error.localizedDescription.UTF8String : "No displays");
                return;
            }
            
            SCDisplay *display = content.displays.firstObject;
            SCStreamConfiguration *config = [[SCStreamConfiguration alloc] init];
            config.capturesAudio = YES;
            config.excludesCurrentProcessAudio = YES;
            config.sampleRate = 44100;
            config.channelCount = 1;
            
            SCContentFilter *filter = [[SCContentFilter alloc] initWithDisplay:display excludingApplications:@[] exceptingWindows:@[]];
            
            globalCapture = [[AudioCapture alloc] init];
            globalCapture.callback = callback;
            
            globalStream = [[SCStream alloc] initWithFilter:filter configuration:config delegate:nil];
            
            NSError *addErr = nil;
            [globalStream addStreamOutput:globalCapture type:SCStreamOutputTypeAudio sampleHandlerQueue:dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0) error:&addErr];
            
            if (addErr) {
                fprintf(stderr, "SCK Add Output Error: %s\n", addErr.localizedDescription.UTF8String);
                return;
            }
            
            [globalStream startCaptureWithCompletionHandler:^(NSError *startErr) {
                if (startErr) {
                    fprintf(stderr, "SCK Start Error: %s\n", startErr.localizedDescription.UTF8String);
                } else {
                    fprintf(stderr, "--- Audio capture started natively via macOS ScreenCaptureKit ---\n");
                }
            }];
        }];
        
        // Start the run loop so this background thread can process SCK callbacks!
        [[NSRunLoop currentRunLoop] run];
    } else {
        fprintf(stderr, "ScreenCaptureKit requires macOS 12.3 or newer.\n");
    }
}

File: agent/src/SCKAudioCapture.swift
import ScreenCaptureKit
import Foundation

// A simple Swift helper to capture system audio via ScreenCaptureKit 
// and write raw PCM data to stdout for the Rust Agent to consume.

class AudioCapture: NSObject, SCStreamOutput {
    func stream(_ stream: SCStream, didOutputSampleBuffer sampleBuffer: CMSampleBuffer, of type: SCStreamOutputType) {
        guard type == .audio else { return }
        guard let blockBuffer = CMSampleBufferGetDataBuffer(sampleBuffer) else { return }
        
        var length = 0
        var ptr: UnsafeMutablePointer<Int8>?
        
        CMBlockBufferGetDataPointer(blockBuffer, atOffset: 0, lengthAtOffsetOut: nil, totalLengthOut: &length, dataPointerOut: &ptr)
        
        if let dataPtr = ptr {
            let data = Data(bytes: dataPtr, count: length)
            FileHandle.standardOutput.write(data)
        }
    }
}

@main
struct App {
    static func main() async {
        do {
            let content = try await SCShareableContent.excludingDesktopWindows(false, onScreenWindowsOnly: true)
            
            // Capture all system audio
            let config = SCStreamConfiguration()
            config.capturesAudio = true
            config.excludesCurrentProcessAudio = true // Don't capture yourself
            
            // Find the main display for the filter
            guard let display = content.displays.first else {
                fputs("No displays found\n", stderr)
                exit(1)
            }
            
            let filter = SCContentFilter(display: display, excludingApplications: [], exceptingWindows: [])
            
            let capture = AudioCapture()
            let stream = SCStream(filter: filter, configuration: config, delegate: nil)
            
            try stream.addStreamOutput(capture, type: .audio, sampleHandlerQueue: .global())
            try await stream.startCapture()
            
            fputs("Audio capture started via ScreenCaptureKit\n", stderr)
            
            // Keep running
            while true {
                try await Task.sleep(nanoseconds: 1_000_000_000)
            }
        } catch {
            fputs("Error starting capture: \(error)\n", stderr)
            exit(1)
        }
    }
}
